dali/test/python/test_operator_subscript.py

```diff
@@ -0,0 +1,140 @@
+    yield _test_indexing, data_gen, "AB", "AB", lambda x: x[-1:,:-2], None
+    yield _test_indexing, data_gen, "AB", "AB", lambda x: x[:-1,:-1], None
+    yield _test_indexing, data_gen, "AB", "B", lambda x: x[1,:2], None
+    yield _test_indexing, data_gen, "AB", "B", lambda x: x[1,:-2], None
+    yield _test_indexing, data_gen, "AB", "A", lambda x: x[:-1,-1], None
+    yield _test_indexing, data_gen, "AB", "A", lambda x: x[:-1,0], None
+
+def test_swapped_ends():
+    data = [np.uint8([1,2,3]),np.uint8([1,2])]
+    src = fn.external_source(lambda: data)
+    pipe = index_pipe(src, lambda x: x[2:1])
+    pipe.build()
+    inp, cpu, gpu = pipe.run()
+    for i in range(len(inp)):
+        x = inp.at(i)
+        assert np.array_equal(x[2:1], cpu.at(i))
+        assert np.array_equal(x[2:1], gpu.as_cpu().at(i))
+
+def test_noop():
+    node = dali.types.Constant(np.float32([1,2,2]))
+    indexed = node[:]
+    assert node is indexed
+
+def test_runtime_indexing():
+    def data_gen():
+        return [np.float32([[0,1,2],[3,4,5]]), np.float32([[0,1],[2,3],[4,5]])]
+    src = fn.external_source(data_gen)
+    lo_idxs = [np.array(x, dtype=np.int64) for x in [1, -5, 0, 2, -2, 1]]
+    hi_idxs = [np.array(x, dtype=np.int16) for x in [5, -1, 1, 2, 4]]
+    lo0 = fn.external_source(source=lo_idxs, batch=False, cycle=True)
+    hi1 = fn.external_source(source=hi_idxs, batch=False, cycle=True)
+    pipe = index_pipe(src, lambda x: x[lo0:, :hi1])
+    pipe.build()
+    j = 0
+    k = 0
+    for _ in range(4):
+        inp, cpu, gpu = pipe.run()
+        for i in range(len(inp)):
+            x = inp.at(i)
+            ref = x[lo_idxs[j]:, :hi_idxs[k]]
+            j = (j + 1) % len(lo_idxs)
+            k = (k + 1) % len(hi_idxs)
+            assert np.array_equal(ref, cpu.at(i))
+            assert np.array_equal(ref, gpu.as_cpu().at(i))
+
+def test_constant_ranges():
+    def data_gen():
+        return [np.float32([[0,1,2],[3,4,5]]), np.float32([[0,1],[2,3],[4,5]])]
+    yield _test_indexing, data_gen, "AB", "",       lambda x: x[1:,dali.newaxis,:2],        lambda x: x[1:,np.newaxis,:2]
+    yield _test_indexing, data_gen, "AB", "CAB",    lambda x: x[dali.newaxis("C"),-1:,:-2], lambda x: x[np.newaxis,-1:,:-2]
+    yield _test_indexing, data_gen, "AB", "ACB",    lambda x: x[:,dali.newaxis("C"),:],     lambda x: x[:,np.newaxis,:]
+    yield _test_indexing, data_gen, "AB", "C",      lambda x: x[1,dali.newaxis("C"),1],     lambda x: x[1,np.newaxis,1]
+
+def _test_inconsistent_args(device, args):
+    data = [np.uint8([[1,2,3]]),np.uint8([[1,2]])]
+    pipe = Pipeline(1, 1, 0)
+    src = fn.external_source(lambda: data, device=device)
+    pipe.set_outputs(fn.tensor_subscript(src, **args))
+    with assert_raises(RuntimeError):
+        pipe.build()
+
+def test_inconsistent_args():
+    for device in ["cpu", "gpu"]:
+        for args in [
+                { "lo_0":0, "at_0":0 },
+                { "at_0":0, "step_0":1 },
```

Comment:
Can you also check step > 1?

Reply:
Well.. I could, but on the other hand, that's "not implemented", not "user error" . I can add it, I guess.

Reply:
Just wanted to make sure that we error as we should.

Reply:
Done in a separate test.